// rhino-editor/cdn/exports/chunks/chunk-QELSW4JY.js
// src/exports/translations.ts
var isiOS = /Mac|iOS|iPhone|iPad|iPod/i.test(
  window.navigator.platform
);
var modifierKey = isiOS ? "cmd" : "ctrl";
var fileUploadErrorMessage = "Unable to upload this file.";
var captionPlaceholder = "Add a caption...";
var translations = {
  attachFiles: "Attach Files",
  bold: `Bold <${modifierKey}+b>`,
  italics: `Italicize <${modifierKey}+i>`,
  strike: `Strikethrough <${modifierKey}+shift+x>`,
  link: `Link <${modifierKey}+k>`,
  heading: `Heading <${modifierKey}+alt+1>`,
  blockQuote: `Blockquote <${modifierKey}+shift+b>`,
  codeBlock: `Codeblock <${modifierKey}+e>`,
  bulletList: `Bullet List <${modifierKey}+shift+7>`,
  orderedList: `Ordered List <${modifierKey}+shift+8>`,
  undo: `Undo <${modifierKey}+z>`,
  redo: `Redo <${modifierKey}+shift+z>`,
  linkDialogLink: "Link",
  linkDialogUnlink: "Unlink",
  placeholder: "Write something...",
  increaseIndentation: "Increase indentation",
  decreaseIndentation: "Decrease indentation",
  fileUploadErrorMessage,
  captionPlaceholder
};

export {
  isiOS,
  modifierKey,
  fileUploadErrorMessage,
  captionPlaceholder,
  translations
};
