// rhino-editor/cdn/exports/chunks/chunk-MDI2LRIY.js
import {
  Node,
  mergeAttributes
} from "./chunk-MF337FR6.js";
import {
  findAttribute
} from "./chunk-4ULJOSG5.js";

// src/exports/extensions/image.ts
var Image = Node.create({
  name: "attachment-image",
  selectable: false,
  draggable: false,
  group: "block",
  addOptions() {
    return {
      HTMLAttributes: {}
    };
  },
  addAttributes() {
    return {
      src: {
        default: "",
        parseHTML: (element) => findAttribute(element, "url")
      },
      height: {
        default: "",
        parseHTML: (element) => findAttribute(element, "height")
      },
      width: {
        default: "",
        parseHTML: (element) => findAttribute(element, "width")
      },
      attachmentId: {
        default: null
      }
    };
  },
  parseHTML() {
    return [
      {
        tag: "figure[data-trix-attachment] img[src]"
      }
    ];
  },
  renderHTML({ HTMLAttributes }) {
    return [
      "img",
      mergeAttributes(this.options.HTMLAttributes, HTMLAttributes)
    ];
  }
});

export {
  Image
};
